Add -Z llvm_module_flag

paulmenage opened the pull request:
Allow adding values to the `!llvm.module.flags` metadata for a generated module.  The syntax is

`-Z llvm_module_flag=<name>:<type>:<value>:<behavior>`

Currently only u32 values are supported but the type is required to be specified for forward compatibility.  The `behavior` element must match one of the named LLVM metadata behaviors.

This flag is expected to be perma-unstable.

paulmenage commented:
r?

rust-log-analyzer commented:
The job **`mingw-check`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/6454022703/job/17518716477) [(plain)](https://github.com/rust-lang/rust/commit/892f482a511419ea7e74821fac349dbd48827acb/checks/17518716477/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
GITHUB_ACTION=__run_7
GITHUB_ACTIONS=true
GITHUB_ACTION_REF=
GITHUB_ACTION_REPOSITORY=
GITHUB_ACTOR=paulmenage
GITHUB_API_URL=https://api.github.com
GITHUB_BASE_REF=master
GITHUB_ENV=/home/runner/work/_temp/_runner_file_commands/set_env_621dd46b-43c7-4615-8660-a8973eaeb63e
GITHUB_EVENT_NAME=pull_request
---
GITHUB_SERVER_URL=https://github.com
GITHUB_SHA=be2a7ff23ecb8e1aabdc9658d45c0f7892fcb031
GITHUB_STATE=/home/runner/work/_temp/_runner_file_commands/save_state_621dd46b-43c7-4615-8660-a8973eaeb63e
GITHUB_STEP_SUMMARY=/home/runner/work/_temp/_runner_file_commands/step_summary_621dd46b-43c7-4615-8660-a8973eaeb63e
GITHUB_TRIGGERING_ACTOR=paulmenage
GITHUB_WORKFLOW_REF=rust-lang/rust/.github/workflows/ci.yml@refs/pull/116555/merge
GITHUB_WORKFLOW_SHA=be2a7ff23ecb8e1aabdc9658d45c0f7892fcb031
GITHUB_WORKSPACE=/home/runner/work/rust/rust
GOROOT_1_19_X64=/opt/hostedtoolcache/go/1.19.13/x64
---
    Checking rustc_mir_transform v0.0.0 (/checkout/compiler/rustc_mir_transform)
    Checking rustc_ty_utils v0.0.0 (/checkout/compiler/rustc_ty_utils)
    Checking rustc_interface v0.0.0 (/checkout/compiler/rustc_interface)
    Checking rustc_driver_impl v0.0.0 (/checkout/compiler/rustc_driver_impl)
error[E0277]: can't compare `Vec<(std::string::String, u32, std::string::String)>` with `std::option::Option<&str>`
    |
86  |   macro_rules! assert_ne {
    |   ---------------------- in this expansion of `assert_ne!` (#2)
...
...
90  |                   if *left_val == *right_val {
    |                                ^^ no implementation for `Vec<(std::string::String, u32, std::string::String)> == std::option::Option<&str>`
   ::: compiler/rustc_interface/src/tests.rs:746:5
    |
746 | /     macro_rules! tracked {
746 | /     macro_rules! tracked {
747 | |         ($name: ident, $non_default_value: expr) => {
748 | |             opts = reference.clone();
749 | |             assert_ne!(opts.unstable_opts.$name, $non_default_value);
...   |
752 | |         };
753 | |     }
    | |_____- in this expansion of `tracked!` (#1)
    | |_____- in this expansion of `tracked!` (#1)
...
795 |       tracked!(llvm_module_flag, Some("bar:u32:123:max"));
    |
    |
    = help: the trait `std::cmp::PartialEq<std::option::Option<&str>>` is not implemented for `Vec<(std::string::String, u32, std::string::String)>`
    = help: the following other types implement trait `std::cmp::PartialEq<Rhs>`:
              <Vec<T, A> as std::cmp::PartialEq<[U; N]>>
              <Vec<T, A> as std::cmp::PartialEq<[U]>>
              <Vec<T, A1> as std::cmp::PartialEq<Vec<U, A2>>>
              <Vec<T, A> as std::cmp::PartialEq<&[U]>>
              <Vec<T, A> as std::cmp::PartialEq<&mut [U]>>
              <Vec<T, A> as std::cmp::PartialEq<&[U; N]>>
error[E0308]: mismatched types
   --> compiler/rustc_interface/src/tests.rs:795:32
    |
    |
750 |             opts.unstable_opts.$name = $non_default_value;
...
...
795 |     tracked!(llvm_module_flag, Some("bar:u32:123:max"));
    |                                ^^^^^^^^^^^^^^^^^^^^^^^ expected `Vec<(String, u32, String)>`, found `Option<&str>`
    |
    = note: expected struct `Vec<(std::string::String, u32, std::string::String)>`

Some errors have detailed explanations: E0277, E0308.
For more information about an error, try `rustc --explain E0277`.
    Checking rustc_driver v0.0.0 (/checkout/compiler/rustc_driver)
```

</details>

rust-log-analyzer commented:
The job **`x86_64-gnu-llvm-15`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/6454424119/job/17519936402) [(plain)](https://github.com/rust-lang/rust/commit/bdfffdce131c77b84e7589f8601c52498b987166/checks/17519936402/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
GITHUB_ACTION=__run_7
GITHUB_ACTIONS=true
GITHUB_ACTION_REF=
GITHUB_ACTION_REPOSITORY=
GITHUB_ACTOR=paulmenage
GITHUB_API_URL=https://api.github.com
GITHUB_BASE_REF=master
GITHUB_ENV=/home/runner/work/_temp/_runner_file_commands/set_env_af64e9a7-b70a-453a-896a-eaf62e33e7f4
GITHUB_EVENT_NAME=pull_request
---
GITHUB_SERVER_URL=https://github.com
GITHUB_SHA=604449fb664432c3cfa271d1986df53f4bd27d33
GITHUB_STATE=/home/runner/work/_temp/_runner_file_commands/save_state_af64e9a7-b70a-453a-896a-eaf62e33e7f4
GITHUB_STEP_SUMMARY=/home/runner/work/_temp/_runner_file_commands/step_summary_af64e9a7-b70a-453a-896a-eaf62e33e7f4
GITHUB_TRIGGERING_ACTOR=paulmenage
GITHUB_WORKFLOW_REF=rust-lang/rust/.github/workflows/ci.yml@refs/pull/116555/merge
GITHUB_WORKFLOW_SHA=604449fb664432c3cfa271d1986df53f4bd27d33
GITHUB_WORKSPACE=/home/runner/work/rust/rust
GOROOT_1_19_X64=/opt/hostedtoolcache/go/1.19.13/x64
---
---- [codegen] tests/codegen/llvm_module_flags.rs stdout ----

error: compilation failed!
status: exit status: 101
command: RUSTC_ICE="0" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/rustc" "/checkout/tests/codegen/llvm_module_flags.rs" "-Zthreads=1" "-Zsimulate-remapped-rust-src-base=/rustc/FAKE_PREFIX" "-Ztranslate-remapped-path-to-local-path=no" "-Z" "ignore-directory-in-diagnostics-source-blocks=/cargo" "--sysroot" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2" "--target=x86_64-unknown-linux-gnu" "-O" "--emit" "llvm-ir" "-C" "prefer-dynamic" "--out-dir" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/codegen/llvm_module_flags" "-A" "internal_features" "-Crpath" "-Cdebuginfo=0" "-Lnative=/checkout/obj/build/x86_64-unknown-linux-gnu/native/rust-test-helpers" "-L" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/codegen/llvm_module_flags/auxiliary" "-Z" "llvm_module_flag=foo:u32:123:error" "-Z" "llvm_module_flag=bar:u32:42:appendunique"
--- stderr -------------------------------
--- stderr -------------------------------
invalid value for 'append'-type module flag (expected a metadata node)
LLVM ERROR: Broken module found, compilation aborted!
------------------------------------------


```

</details>

paulmenage committed:
Add -Z llvm_module_flag

Allow adding values to the `!llvm.module.flags` metadata for a generated
module.  The syntax is

`-Z llvm_module_flag=<name>:<type>:<value>:<behavior>`

Currently only u32 values are supported but the type is required to be
specified for forward compatibility.  The `behavior` element must match
one of the named LLVM metadata behaviors.viors.

This flag is expected to be perma-unstable.

wesleywiser commented on src/doc/unstable-book/src/compiler-flags/llvm-module-flag.md:
```diff
@@ -0,0 +1,12 @@
+
+`-Z llvm_module_flag=<name>:<type>:<value>:<behavior>`
+
+Currently only u32 values are supported but the type is required to be specified
+for forward compatibility.  The `behavior` element must match one of the named
+LLVM [metadata behaviors](https://llvm.org/docs/LangRef.html#module-flags-metadata)
```
It might be good to have a practical example of why you might want to use this flag here.

dreiss replied:
For setting the small data limit to 0 to prevent the use of .sdata (the purpose of this PR), the flag can be specified as `-Zllvm_module_flag=SmallDataLimit:u32:0:Error`

wesleywiser reviewed (commented):
For `rustc`, perma-unstable compiler flags tend to mostly be used by people working on the compiler itself rather than features users might want to rely on. Could you explain a bit why you want to adjust LLVM module flags in this way? 

Thanks! 🙂

paulmenage commented:
My immediate motivation is that I'm working on an embedded system where I really need LLVM to not put symbols into the "small data segment" (`.sdata` and `.sbss`) but instead put each symbol in its own section as normal.

The platforms where LLVM supports small data segments each have their own way of controlling the threshold for what might get put in the small data segments. Three of them (MIPS, M68K, Hexagon) use LLVM arguments (although different args for each platform) so if I was targeting those platforms the existing `-C llvm-args` flag would be enough. RiscV (which is the platform I actually care about) uses an LLVM module flag (`SmallDataLimit`). So being able to set module flags would solve the problem here, in the same way that you can set LLVM arguments currently.

Possibly the better solution to the underlying problem would be a rustc flag, something like `-C small-data-threshold`, as in https://github.com/rust-lang/rust/compare/master...paulmenage:rust:small-data-limit.

But,
- that seems rather more specific to my needs rather than generically useful
- the tests to prove that it works are painful since it seems that some of the platforms are kinda non-deterministic in when they actually put things in the small data segments, regardless of what LLVM args are passed

So I figured that a module-flags analogue of the existing `-C llvm-args` option would fit in better with the existing compiler, and be less controversial.

bors commented:
:umbrella: The latest upstream changes (presumably #115214) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

bors commented:
:umbrella: The latest upstream changes (presumably #117706) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

wesleywiser commented:
Thanks @paulmenage! I agree, since we already have `-C llvm-args`, it makes sense to allow setting LLVM module flags as well. 

@bors r+ rollup

bors commented:
:pushpin: Commit 2e6b57541dc28cdd954b1ce2300b88ccc5c29d5c has been approved by `wesleywiser`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=wesleywiser 2e6b57541dc28cdd954b1ce2300b88ccc5c29d5c -->
<!-- homu: {"type":"Approved","sha":"2e6b57541dc28cdd954b1ce2300b88ccc5c29d5c","approver":"wesleywiser","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->

bors commented:
:hourglass: Testing commit 2e6b57541dc28cdd954b1ce2300b88ccc5c29d5c with merge 0b244796388a929affbea546f1c886dc2e482898...
<!-- homu: {"type":"BuildStarted","head_sha":"2e6b57541dc28cdd954b1ce2300b88ccc5c29d5c","merge_sha":"0b244796388a929affbea546f1c886dc2e482898"} -->

bors commented:
:sunny: Test successful - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/6880415485/job/18718406152)
Approved by: wesleywiser
Pushing 0b244796388a929affbea546f1c886dc2e482898 to master...
<!-- homu: {"type":"BuildCompleted","approved_by":"wesleywiser","base_ref":"master","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/actions/runs/6880415485/job/18718406152"},"merge_sha":"0b244796388a929affbea546f1c886dc2e482898"} -->

rust-timer commented:
Finished benchmarking commit (0b244796388a929affbea546f1c886dc2e482898): [comparison URL](https://perf.rust-lang.org/compare.html?start=d4559c01c2bdebad53c01f85064617ac61563a38&end=0b244796388a929affbea546f1c886dc2e482898&stat=instructions:u).

### Overall result: no relevant changes - no action needed


@rustbot label: -perf-regression



### [Instruction count](https://perf.rust-lang.org/compare.html?start=d4559c01c2bdebad53c01f85064617ac61563a38&end=0b244796388a929affbea546f1c886dc2e482898&stat=instructions:u)
This benchmark run did not return any relevant results for this metric.

### [Max RSS (memory usage)](https://perf.rust-lang.org/compare.html?start=d4559c01c2bdebad53c01f85064617ac61563a38&end=0b244796388a929affbea546f1c886dc2e482898&stat=max-rss)
<details>
<summary>Results</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|                                    | mean | range        | count |
|:----------------------------------:|:----:|:------------:|:-----:|
| Regressions ❌ <br /> (primary)    | 0.7% | [0.7%, 0.7%] | 1     |
| Regressions ❌ <br /> (secondary)  | 2.2% | [2.2%, 2.2%] | 1     |
| Improvements ✅ <br /> (primary)   | -    | -            | 0     |
| Improvements ✅ <br /> (secondary) | -    | -            | 0     |
| All ❌✅ (primary)                 | 0.7% | [0.7%, 0.7%] | 1     |
</details>

### [Cycles](https://perf.rust-lang.org/compare.html?start=d4559c01c2bdebad53c01f85064617ac61563a38&end=0b244796388a929affbea546f1c886dc2e482898&stat=cycles:u)
<details>
<summary>Results</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | 0.5%  | [0.5%, 0.5%]   | 1     |
| Regressions ❌ <br /> (secondary)  | -     | -              | 0     |
| Improvements ✅ <br /> (primary)   | -0.5% | [-0.5%, -0.5%] | 1     |
| Improvements ✅ <br /> (secondary) | -     | -              | 0     |
| All ❌✅ (primary)                 | -0.0% | [-0.5%, 0.5%]  | 2     |
</details>

### [Binary size](https://perf.rust-lang.org/compare.html?start=d4559c01c2bdebad53c01f85064617ac61563a38&end=0b244796388a929affbea546f1c886dc2e482898&stat=size:linked_artifact)
This benchmark run did not return any relevant results for this metric.

**Bootstrap**: 674.873s -> 675.256s (0.06%)
**Artifact size**: 311.07 MiB -> 311.08 MiB (0.00%)
